File: py/listenfree_sdk.py
# Listenfree SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import ListenfreeUtility
from core.spec import ListenfreeSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import ListenfreeBaseFeature
from features import _make_feature


class ListenfreeSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = ListenfreeUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return ListenfreeUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = ListenfreeSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def listening_room(self):
        """Idiomatic facade: client.listening_room.list() / client.listening_room.load({"id": ...})."""
        from entity.listening_room_entity import ListeningRoomEntity
        cached = getattr(self, "_listening_room", None)
        if cached is None:
            cached = ListeningRoomEntity(self, None)
            self._listening_room = cached
        return cached

    def ListeningRoom(self, data=None):
        # Deprecated: use client.listening_room instead.
        from entity.listening_room_entity import ListeningRoomEntity
        return ListeningRoomEntity(self, data)


    @property
    def music(self):
        """Idiomatic facade: client.music.list() / client.music.load({"id": ...})."""
        from entity.music_entity import MusicEntity
        cached = getattr(self, "_music", None)
        if cached is None:
            cached = MusicEntity(self, None)
            self._music = cached
        return cached

    def Music(self, data=None):
        # Deprecated: use client.music instead.
        from entity.music_entity import MusicEntity
        return MusicEntity(self, data)


    @property
    def offline_download(self):
        """Idiomatic facade: client.offline_download.list() / client.offline_download.load({"id": ...})."""
        from entity.offline_download_entity import OfflineDownloadEntity
        cached = getattr(self, "_offline_download", None)
        if cached is None:
            cached = OfflineDownloadEntity(self, None)
            self._offline_download = cached
        return cached

    def OfflineDownload(self, data=None):
        # Deprecated: use client.offline_download instead.
        from entity.offline_download_entity import OfflineDownloadEntity
        return OfflineDownloadEntity(self, data)


    @property
    def playlist(self):
        """Idiomatic facade: client.playlist.list() / client.playlist.load({"id": ...})."""
        from entity.playlist_entity import PlaylistEntity
        cached = getattr(self, "_playlist", None)
        if cached is None:
            cached = PlaylistEntity(self, None)
            self._playlist = cached
        return cached

    def Playlist(self, data=None):
        # Deprecated: use client.playlist instead.
        from entity.playlist_entity import PlaylistEntity
        return PlaylistEntity(self, data)


    @property
    def search(self):
        """Idiomatic facade: client.search.list() / client.search.load({"id": ...})."""
        from entity.search_entity import SearchEntity
        cached = getattr(self, "_search", None)
        if cached is None:
            cached = SearchEntity(self, None)
            self._search = cached
        return cached

    def Search(self, data=None):
        # Deprecated: use client.search instead.
        from entity.search_entity import SearchEntity
        return SearchEntity(self, data)


    @property
    def song(self):
        """Idiomatic facade: client.song.list() / client.song.load({"id": ...})."""
        from entity.song_entity import SongEntity
        cached = getattr(self, "_song", None)
        if cached is None:
            cached = SongEntity(self, None)
            self._song = cached
        return cached

    def Song(self, data=None):
        # Deprecated: use client.song instead.
        from entity.song_entity import SongEntity
        return SongEntity(self, data)


    @property
    def stream(self):
        """Idiomatic facade: client.stream.list() / client.stream.load({"id": ...})."""
        from entity.stream_entity import StreamEntity
        cached = getattr(self, "_stream", None)
        if cached is None:
            cached = StreamEntity(self, None)
            self._stream = cached
        return cached

    def Stream(self, data=None):
        # Deprecated: use client.stream instead.
        from entity.stream_entity import StreamEntity
        return StreamEntity(self, data)


    @property
    def video(self):
        """Idiomatic facade: client.video.list() / client.video.load({"id": ...})."""
        from entity.video_entity import VideoEntity
        cached = getattr(self, "_video", None)
        if cached is None:
            cached = VideoEntity(self, None)
            self._video = cached
        return cached

    def Video(self, data=None):
        # Deprecated: use client.video instead.
        from entity.video_entity import VideoEntity
        return VideoEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk

File: lua/listenfree_sdk.lua
-- Listenfree SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local ListenfreeSDK = {}
ListenfreeSDK.__index = ListenfreeSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

ListenfreeSDK._make_feature = _make_feature


function ListenfreeSDK.new(options)
  local self = setmetatable({}, ListenfreeSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features from config.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local feature_items = vs.items(feature_opts)
    if feature_items ~= nil then
      for _, item in ipairs(feature_items) do
        local fname = item[1]
        local fopts = helpers.to_map(item[2])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function ListenfreeSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function ListenfreeSDK:get_utility()
  return Utility.copy(self._utility)
end


function ListenfreeSDK:get_root_ctx()
  return self._rootctx
end


function ListenfreeSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function ListenfreeSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:listening_room():list() / client:listening_room():load({ id = ... })
function ListenfreeSDK:listening_room(data)
  local EntityMod = require("entity.listening_room_entity")
  if data == nil then
    if self._listening_room == nil then
      self._listening_room = EntityMod.new(self, nil)
    end
    return self._listening_room
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:listening_room() instead.
function ListenfreeSDK:ListeningRoom(data)
  local EntityMod = require("entity.listening_room_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:music():list() / client:music():load({ id = ... })
function ListenfreeSDK:music(data)
  local EntityMod = require("entity.music_entity")
  if data == nil then
    if self._music == nil then
      self._music = EntityMod.new(self, nil)
    end
    return self._music
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:music() instead.
function ListenfreeSDK:Music(data)
  local EntityMod = require("entity.music_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:offline_download():list() / client:offline_download():load({ id = ... })
function ListenfreeSDK:offline_download(data)
  local EntityMod = require("entity.offline_download_entity")
  if data == nil then
    if self._offline_download == nil then
      self._offline_download = EntityMod.new(self, nil)
    end
    return self._offline_download
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:offline_download() instead.
function ListenfreeSDK:OfflineDownload(data)
  local EntityMod = require("entity.offline_download_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:playlist():list() / client:playlist():load({ id = ... })
function ListenfreeSDK:playlist(data)
  local EntityMod = require("entity.playlist_entity")
  if data == nil then
    if self._playlist == nil then
      self._playlist = EntityMod.new(self, nil)
    end
    return self._playlist
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:playlist() instead.
function ListenfreeSDK:Playlist(data)
  local EntityMod = require("entity.playlist_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:search():list() / client:search():load({ id = ... })
function ListenfreeSDK:search(data)
  local EntityMod = require("entity.search_entity")
  if data == nil then
    if self._search == nil then
      self._search = EntityMod.new(self, nil)
    end
    return self._search
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:search() instead.
function ListenfreeSDK:Search(data)
  local EntityMod = require("entity.search_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:song():list() / client:song():load({ id = ... })
function ListenfreeSDK:song(data)
  local EntityMod = require("entity.song_entity")
  if data == nil then
    if self._song == nil then
      self._song = EntityMod.new(self, nil)
    end
    return self._song
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:song() instead.
function ListenfreeSDK:Song(data)
  local EntityMod = require("entity.song_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:stream():list() / client:stream():load({ id = ... })
function ListenfreeSDK:stream(data)
  local EntityMod = require("entity.stream_entity")
  if data == nil then
    if self._stream == nil then
      self._stream = EntityMod.new(self, nil)
    end
    return self._stream
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:stream() instead.
function ListenfreeSDK:Stream(data)
  local EntityMod = require("entity.stream_entity")
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:video():list() / client:video():load({ id = ... })
function ListenfreeSDK:video(data)
  local EntityMod = require("entity.video_entity")
  if data == nil then
    if self._video == nil then
      self._video = EntityMod.new(self, nil)
    end
    return self._video
  end
  return EntityMod.new(self, data)
end

-- Deprecated: use client:video() instead.
function ListenfreeSDK:Video(data)
  local EntityMod = require("entity.video_entity")
  return EntityMod.new(self, data)
end




function ListenfreeSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = ListenfreeSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return ListenfreeSDK
